compiler/dyno/lib/resolution/can-pass.cpp

```diff
@@ -977,6 +977,200 @@ CanPassResult CanPassResult::canPass(Context* context,
+  bool isRef = false;
+  bool isType = false;
+  bool isParam = false;
+  bool isValid = false;
+
+  KindProperties() {}
+
+  KindProperties(bool isConst, bool isRef, bool isType,
+                   bool isParam)
+    : isConst(isConst), isRef(isRef), isType(isType),
+      isParam(isParam), isValid(true) {}
+
+ public:
+  static KindProperties fromKind(QualifiedType::Kind kind) {
+    if (kind == QualifiedType::TYPE)
+      return KindProperties(false, false, true, false);
+    if (kind == QualifiedType::PARAM)
+      // Mark params as const to cover the case in which a
+      // param decays to a const var.
+      return KindProperties(true, false, false, true);
+    auto isConst = isConstIntent(kind);
+    auto isRef = isRefIntent(kind);
+    return KindProperties(isConst, isRef, false, false);
+  }
+
+  void invalidate() {
+    isRef = isConst = isType = isParam = isValid = false;
+  }
+
+  void setParam(bool isParam) {
+    this->isParam = isParam;
+  }
+
+  void combineWith(const KindProperties& other) {
+    if (!isValid) return;
+    if (!other.isValid ||
+        (isType ^ other.isType)) {
+      // Can't mix types and non-types.
+      invalidate();
+      return;
+    }
+    isConst = isConst || other.isConst;
+    isRef = isRef && other.isRef;
+    isParam = isParam && other.isParam;
+  }
+
+  void strictCombineWith(const KindProperties& other) {
+    if (!isValid) return;
+    if (!other.isValid ||
+        (isType ^ other.isType)) {
+      // Can't mix types and non-types.
+      invalidate();
+      return;
+    }
+    if (isParam && !other.isParam) {
+      // If a param is required, can't return a non-param.
+      invalidate();
+      return;
+    }
+    // Ensuring that everything can actually be made
+    // into a reference and const will happen later.
+    // leave isRef and isConst as specified.
+    // We could do some checking now, but that might be a bit premature.
+  }
+
+  QualifiedType::Kind toKind() const {
+    if (!isValid) return QualifiedType::UNKNOWN;
+    if (isType) return QualifiedType::TYPE;
+    if (isParam) return QualifiedType::PARAM;
+    if (isConst) {
+      return isRef ? QualifiedType::CONST_REF : QualifiedType::CONST_VAR ;
+    } else {
+      return isRef ? QualifiedType::REF : QualifiedType::VAR ;
+    }
+  }
+
+  bool valid() const { return isValid; }
+};
+
+static QualifiedType findByPassing(Context* context,
+                                   const std::vector<QualifiedType>& types) {
+  for (auto& type : types) {
+    bool fitsOthers = true;
+    for (auto& otherType : types) {
+      if (otherType == type) continue;
+      auto passes = canPass(context, otherType, type);
+      if (!passes.passes() || passes.promotes()) {
+        fitsOthers = false;
+        break;
+      }
+    }
+    if (fitsOthers) return type;
+  }
+  return QualifiedType();
+}
+
+QualifiedType commonType(Context* context,
+                                const std::vector<QualifiedType>& types,
```

Comment:
indenting seems weird. can you remove some spaces from this line and the next two so they line up with `Context` ?